wasi/src/main/java/com/dylibso/chicory/wasi/WasiPreview1.java

```diff
@@ -1339,9 +1338,122 @@ public int pathUnlinkFile(int dirFd, @Buffer String rawPath) {
+        }
+
+        int nevents = 0;
+        List<Entry<Long, Long>> clockEvents = new ArrayList<>();
+        List<Entry<InStream, Long>> readEvents = new ArrayList<>();
+
+        for (int i = 0; i < nsubscriptions; i++) {
+            long userData = memory.readLong(inPtr);
+            byte eventType = memory.read(inPtr + 8);
+            inPtr += 16;
+            switch (eventType) {
+                case WasiEventType.CLOCK:
+                    int clockId = memory.readInt(inPtr);
+                    long timeout = memory.readLong(inPtr + 8);
+                    short flags = memory.readShort(inPtr + 24);
+                    if (clockId != WasiClockId.REALTIME && clockId != WasiClockId.MONOTONIC) {
+                        return wasiResult(WasiErrno.EINVAL);
+                    }
+                    if (flagSet(flags, WasiSubClockFlags.SUBSCRIPTION_CLOCK_ABSTIME)) {
+                        timeout -= clockTime(clockId);
+                    }
+                    clockEvents.add(Map.entry(timeout, userData));
+                    break;
+                case WasiEventType.FD_READ:
+                case WasiEventType.FD_WRITE:
+                    int fd = memory.readInt(inPtr);
+                    if (fd < 0) {
+                        return wasiResult(WasiErrno.EBADF);
+                    }
+                    var descriptor = descriptors.get(fd);
+                    if (descriptor instanceof InStream && eventType == WasiEventType.FD_READ) {
+                        readEvents.add(Map.entry((InStream) descriptor, userData));
+                    } else if (descriptor instanceof OutStream
+                            && eventType == WasiEventType.FD_WRITE) {
+                        // assume output streams are always writable
+                        writeEvent(memory, outPtr, userData, eventType, WasiErrno.ESUCCESS);
+                        outPtr += 32;
+                        nevents++;
+                    } else {
+                        WasiErrno errno;
+                        if (descriptor == null) {
+                            errno = WasiErrno.EBADF;
+                        } else if (descriptor instanceof OpenFile) {
+                            // per specification: this event always triggers for regular files
+                            errno = WasiErrno.ESUCCESS;
+                        } else {
+                            errno = WasiErrno.ENOTSUP;
+                        }
+                        writeEvent(memory, outPtr, userData, eventType, errno);
+                        outPtr += 32;
+                        nevents++;
+                    }
+                    break;
+                default:
+                    return wasiResult(WasiErrno.EINVAL);
+            }
+            inPtr += 32;
+        }
+
+        long minTimeout = clockEvents.stream().mapToLong(Entry::getKey).min().orElse(0);
+        long start = System.nanoTime();
+        while (true) {
```

Comment:
A `while (true)` with so much code afterwords make me slightly uncomfortable.

In this case, if I'm not missing an invariant:

- `nsubscriptions` == 0
- `readEvents` is an empty List
- `clockEvents` is an empty List
- the loop doesn't end

is it the expected behavior?

Reply:
+1 I think just a few test cases would be useful to clarify the behavior :) 

there's a few here 
- unit tests https://github.com/tetratelabs/wazero/blob/610c202ec48f3a7c729f2bf11707330127ab3689/imports/wasi_snapshot_preview1/poll_test.go
- integration:
  - https://github.com/tetratelabs/wazero/blob/610c202ec48f3a7c729f2bf11707330127ab3689/imports/wasi_snapshot_preview1/wasi_stdlib_test.go#L275
  - https://github.com/tetratelabs/wazero/blob/610c202ec48f3a7c729f2bf11707330127ab3689/imports/wasi_snapshot_preview1/wasi_stdlib_test.go#L501

(it doesn't need to be all, because this is not really covering all of them, just a source of inspiration)

Reply:
You're right, I had a typo on the check at the start. It should be

> If `nsubscriptions` is 0, returns `errno::inval`.

So we will never enter the loop in that case.

Reply:
Thanks for updating this, now it looks safer, in all honesty, I still need to squint a lot my eyes to understand what's going on.

Adding unit tests for the implementation is an option to better clarify the expected behavior, does it sound feasible without a major amount of additional work?

Reply:
I added tests for all cases except for invalid input and `IOException`, and verified this with code coverage in the IDE. I also added comments which hopefully makes the logic easier to follow.